### PR TITLE
CartesianGrid reads axes from context instead of props

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

Almost there and then I can move on to next component

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

No more element cloning

## How Has This Been Tested?

npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [X] All new and existing tests passed.

### CODE REVIEW

File: src/context/chartLayoutContext.tsx [hunk 1]
```diff
@@ -106,10 +109,50 @@ export const useXAxisOrThrow = (xAxisId: string | number): XAxisProps => {
...
+  const yAxisMap = useContext(YAxisContext);
+  const yAxisWithFiniteDomain = find(yAxisMap, axis => every(axis.domain, Number.isFinite));
+  return yAxisWithFiniteDomain || getAnyElementOfObject(yAxisMap);
+};
+
 /**
  * This either finds and returns Axis by the specified ID, or throws an exception if an axis with this ID does not exist.
  *
- * @param yAxisId identifier of the axis - it's either autogenerated ('0'), or passed via `id` prop as <YAxis id='foo' />
+ * @param yAxisId identifier of the axis - it's either autogenerated ('0'), or passed via `id` prop as <Axis id='foo' />
```
Comment (ckifer) [inline src/context/chartLayoutContext.tsx:155]: nit: should this still be YAxis?

File: src/cartesian/CartesianGrid.tsx [hunk 2]
```diff
@@ -371,7 +398,12 @@ export function CartesianGrid(props: Props) {
     height: isNumber(props.height) ? props.height : offset.height,
   };
 
-  const { x, y, width, height, xAxis, yAxis, syncWithTicks, horizontalValues, verticalValues } = propsIncludingDefaults;
+  const { x, y, width, height, syncWithTicks, horizontalValues, verticalValues } = propsIncludingDefaults;
+
+  // @ts-expect-error the scale prop is mixed up - we need to untagle this at some point
```
Comment (ckifer) [inline src/cartesian/CartesianGrid.tsx:403]: Hmm https://github.com/recharts/recharts/blob/master/src/util/types.ts#L26

We definitely aren't using correct scale types lol

**Review by ckifer — APPROVED** [review]
LGTM

File: src/context/chartLayoutContext.tsx (hunk 1 above)
Comment (PavelVanecek) [reply src/context/chartLayoutContext.tsx:155]:
```suggestion
 * @param yAxisId identifier of the axis - it's either autogenerated ('0'), or passed via `id` prop as <YAxis id='foo' />
```

indeed

File: src/cartesian/CartesianGrid.tsx (hunk 2 above)
Comment (PavelVanecek) [reply src/cartesian/CartesianGrid.tsx:403]: Yep, Recharts will accept `scale` externally as a string, and then parse it and evaluate as one of the d3 scales - but the types for the "external scale" and "internal scale" are mixed. It's on my list.

**Comment by ckifer** [comment]
🚢

**Pull request merged**